calico_node/readiness/bird/bird.go

```diff
@@ -0,0 +1,275 @@
+	c, err := net.Dial("unix", fmt.Sprintf("/var/run/calico/bird%s.ctl", birdSuffix))
+	if err != nil {
+		// If that fails, try connecting to bird socket in `/var/run/bird` (which is the
+		// default socket location for bird install) for non-containerized installs
+		log.Debugln("Failed to connect to BIRD socket in /var/run/calico, trying /var/run/bird")
+		c, err = net.Dial("unix", fmt.Sprintf("/var/run/bird/bird%s.ctl", birdSuffix))
+		if err != nil {
+			return fmt.Errorf("Error querying BIRD: unable to connect to BIRDv%s socket: %v", ipv, err)
+		}
+	}
+	defer c.Close()
+
+	// To query the current state of the BGP peers, we connect to the BIRD
+	// socket and send a "show protocols" message.  BIRD responds with
+	// peer data in a table format.
+	//
+	// Send the request.
+	_, err = c.Write([]byte("show status\n"))
+	if err != nil {
+		return fmt.Errorf("Error executing command: unable to write to BIRD socket: %s", err)
+	}
+
+	// The following is sample output from BIRD
+	//
+	// 0001 BIRD v0.3.2+birdv1.6.3 ready.
+	// 1000-BIRD v0.3.2+birdv1.6.3
+	// 1011-Router ID is 172.18.0.4
+	// 	 Current server time is 2018-04-03 22:59:51
+	// 	 Last reboot on 2018-04-03 22:59:26
+	// 	 Last reconfiguration on 2018-04-03 22:59:26
+	// 0024-Graceful restart recovery in progress
+	// 	   Waiting for 1 protocols to recover
+	// 	   Wait timer is 215/240
+	// 0013 Daemon is up and running
+	scanner := bufio.NewScanner(c)
+
+	// Set a time-out for reading from the socket connection.
+	c.SetReadDeadline(time.Now().Add(birdTimeOut))
+
+	for scanner.Scan() {
+		// Process the next line that has been read by the scanner.
+		str := scanner.Text()
+		log.Debugf("Read: %s\n", str)
+		if strings.HasPrefix(str, "0013") {
+			// "0013" code is the final line and indicates that BIRD is ready,
+			break
+		} else if strings.HasPrefix(str, "0001") {
+			// "0001" code means BIRD is ready.
+		} else if strings.HasPrefix(str, "1000") {
+			// "1000" code shows the bird version
+		} else if strings.HasPrefix(str, "1011") {
+			// "1011" shows uptime
+		} else if strings.HasPrefix(str, "0024") {
+			// "0024" code indicates the start of a graceful restart status report.
+			// This means a GR is in progress.
+			return errors.New("graceful restart in progress")
+		} else if strings.HasPrefix(str, " ") {
+			// Row starting with a " " is another row of data.
+		} else {
+			// Format of row is unexpected.
+			return errors.New("unexpected output line from BIRD")
+		}
+
+		// Before reading the next line, adjust the time-out for
+		// reading from the socket connection.
+		c.SetReadDeadline(time.Now().Add(birdTimeOut))
+	}
+
+	return scanner.Err()
+}
+
+// bgpPeer is a structure containing details about a BGP peer.
+type bgpPeer struct {
+	PeerIP   string
+	PeerType string
+	State    string
+	Since    string
+	BGPState string
+	Info     string
+}
+
+func GetBirdPeers(ipv string) ([]bgpPeer, error) {
+	log.Debugf("Print BIRD peers for IPv%s", ipv)
+	birdSuffix := ""
```

Comment:
I know this is copy-paste, but feels cleaner if we just subbed the whole socket here. e.g.

```go
sck := "/var/run/bird/bird.ctl"
if ipv == "6" {
  sck = "/var/run/bird/bird6.ctl"
}
```

Reply:
Going to leave this as is for now, as we also need to template the backup socket location (/var/run/calico/bird.ctl)